File: retort-annotations/src/main/lombok/co/kaioru/retort/annotation/adapter/CommandAnnotationGenerator.java
package co.kaioru.retort.annotation.adapter;

import co.kaioru.retort.AbstractCommand;
import co.kaioru.retort.ICommand;
import co.kaioru.retort.ICommandContext;
import co.kaioru.retort.annotation.AbstractAnnotationGenerator;
import co.kaioru.retort.annotation.type.Command;
import co.kaioru.retort.exception.CommandException;
import co.kaioru.retort.exception.CommandReflectionException;
import lombok.Getter;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

@Getter
public class CommandAnnotationGenerator<I extends ICommandContext, O> extends AbstractAnnotationGenerator<I, O> {

    private final Class<I> inputClass;
    private final Class<O> outputClass;

    public CommandAnnotationGenerator(Class<I> inputClass, Class<O> outputClass) {
        this.inputClass = inputClass;
        this.outputClass = outputClass;
    }

    @SuppressWarnings("unchecked")
    @Override
    public ICommand<I, O> generate(Object object, Method method) {
        ICommand<I, O> command = null;

        if (method.isAnnotationPresent(Command.class)) {
            if (method.getParameterCount() == 1
                    && method.getParameters()[0].getType().equals(getInputClass())
                    && method.getReturnType().equals(getOutputClass())) {
                String commandName = method.getAnnotation(Command.class).value();

                command = new AbstractCommand<I, O>(commandName) {

                    @Override
                    public O execute(I i) throws CommandException {
                        try {
                            method.setAccessible(true);
                            return (O) method.invoke(object, i);
                        } catch (InvocationTargetException | IllegalAccessException e) {
                            throw new CommandReflectionException();
                        }
                    }

                };
            }
        }
        return command;
    }

}

File: retort-annotations/src/main/kotlin/co/kaioru/retort/annotation/impl/adapter/CommandAnnotationGenerator.kt
package co.kaioru.retort.annotation.impl.adapter

import co.kaioru.retort.ICommand
import co.kaioru.retort.ICommandExecutable
import co.kaioru.retort.annotation.exceptions.AnnotationAdapterException
import co.kaioru.retort.annotation.impl.AnnotationGenerator
import co.kaioru.retort.annotation.type.Command
import co.kaioru.retort.impl.CommandContext
import co.kaioru.retort.inline.command
import java.lang.reflect.Method


class CommandAnnotationGenerator<I : CommandContext, O>(
        val inputClass: Class<I>,
        val outputClass: Class<O>
) : AnnotationGenerator<I, O>() {
    @Suppress("UNCHECKED_CAST")
    override fun generate(any: Any, method: Method): ICommand<I, O> {
        if (method.isAnnotationPresent(Command::class.java)) {
            if (method.parameters.size == 1
                    && method.parameters.first().type == inputClass
                    && method.returnType == outputClass) {
                val name: String = method.getAnnotation(Command::class.java).value

                return command(name) {
                    build(object : ICommandExecutable<I, O> {
                        override fun execute(input: I): O {
                            method.isAccessible = true
                            return method.invoke(any, input) as O
                        }
                    })
                }
            }
        }
        throw AnnotationAdapterException()
    }
}